main.rb

```diff
@@ -0,0 +1,142 @@
+# rubocop: disable Metrics/ModuleLength
+# rubocop: disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+
+module Enumerable
+  def my_each
+    return to_enum :my_each unless block_given?
+
+    index = 0
+    while index < size
+      if is_a? Array
+        yield self[index]
+      elsif is_a? Hash
+        yield keys[index], self[keys[index]]
+      elsif is_a? Range
+        yield to_a[index]
+      end
+      index += 1
+    end
+  end
+
+  def my_each_with_index
+    return to_enum :my_each unless block_given?
+
+    index = 0
+    while index < size
+      if is_a? Array
+        yield self[index], index
+      elsif is_a? Hash
+        yield keys[index], self[keys[index]]
+      elsif is_a? Range
+        yield to_a[index], index
+      end
+      index += 1
+    end
+  end
+
+  def my_select
+    return to_enum :my_select unless block_given?
+
+    if is_a? Array
+      results = []
+      my_each { |x| results << x if yield x }
+    else
+      results = {}
+      my_each { |k, v| results[k] = v if yield k, v }
+    end
+    results
+  end
+
+  def my_any?(args = nil)
+    block_is_true = false
+    if block_given?
+      my_each { |x| block_is_true = true if yield(x) }
+    elsif args.is_a? Regexp
+      my_each { |x| block_is_true = true if x.match(args) }
+    elsif args.is_a? Module
+      my_each { |x| block_is_true = true if x.is_a?(args) }
+    elsif args.nil?
+      my_each { |x| block_is_true = true unless x.nil? || x == false }
+    else
+      my_each { |x| block_is_true = true if x.nil? || x == false }
+    end
+    block_is_true
+  end
+
+  def my_map(&proc)
+    return to_enum :my_map unless block_given?
+
+    results = []
+    my_each { |x| results << proc.call(x) }
+    results
+  end
```

Comment:
[OPTIONAL]

- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend working on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```